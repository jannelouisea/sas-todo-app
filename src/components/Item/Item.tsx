import Card from '@mui/joy/Card';
import Typography from '@mui/joy/Typography';
import PropTypes from 'prop-types';
import moment from 'moment';

import { useAppDispatch, useAppSelector } from 'src/app/hooks'
import { todoItemToggled, todoItemsSorted } from 'src/features/todoItems/todoItemsSlice'
import { isOlderItem as isOlderItemUtil } from 'src/utils';
import HourglassBottomIcon from '@mui/icons-material/HourglassBottom';
import classNames from 'classnames';

import {
    IToDoItem
} from 'src/interfaces/'
import {
    RemoveItemButtton,
    ToggleItemButton
} from 'src/components'
import {
    MUIColor,
    MUIVariant,
    MUISize
} from 'src/static/enums'

type Props = {
    item: IToDoItem
}

function Item({ item }: Props) {
    const dispatch = useAppDispatch();

    const sortBy = useAppSelector(state => state.todoItemsFilters.sortBy);

    const isOlderItem: boolean = isOlderItemUtil(item) && !item.completed;
    const color: MUIColor = item.completed ? MUIColor.Success : isOlderItem ? MUIColor.Warning : MUIColor.Primary;
    const cardVariant: MUIVariant = item.completed ? MUIVariant.Soft : MUIVariant.Outlined;
    const buttonVariant: MUIVariant = item.completed ? MUIVariant.Solid : MUIVariant.Outlined;

    const isOlderIconClassNames = classNames('mr-1', { 'hidden': item.completed || (!isOlderItem) });

    const onToggleItem = (id: string) => {
        dispatch(todoItemToggled(id));
        dispatch(todoItemsSorted(sortBy));
    }

    return (
        <Card
            className='drop-shadow-lg'
            color={color}
            variant={cardVariant}
            size={MUISize.Small}
            sx={{
                borderRadius: '2rem',
                border: isOlderItem ? `1px solid #D4A72C` : ''
            }}
        >
            <div className="flex items-center group">
                <div className="ml-1 mr-4">
                    <ToggleItemButton
                        color={color}
                        variant={buttonVariant}
                        onClick={() => onToggleItem(item.id)}
                    />
                </div>
                <div className="flex-1">
                    <span>{item.text}</span>
                    <div className='flex items-center'>
                        <div className={isOlderIconClassNames}>
                            <HourglassBottomIcon sx={{ fontSize: '1rem' }} />
                        </div>
                        <div className='ml-1'>
                            <Typography
                                textColor={item.completed ? 'neutral.700' : 'neutral.500'}
                                level='body3'
                                variant={MUIVariant.Plain}
                                sx={{ padding: 0, paddingTop: '.25rem' }}
                            >
                                {`Created ${moment(item.createdAt).format('lll')}`}
                            </Typography>
                        </div>
                    </div>
                </div>
                <div className="ml-4 mr-2 invisible group-hover:visible">
                    <RemoveItemButtton itemId={item.id} color={color} />
                </div>
            </div>
        </Card>
    );
}

Item.propTypes = {
    onClick: PropTypes.object
}

export default Item;