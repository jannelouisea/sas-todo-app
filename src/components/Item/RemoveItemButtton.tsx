import IconButton from '@mui/joy/IconButton';
import Delete from '@mui/icons-material/Delete';

import { useAppDispatch } from 'src/app/hooks';
import { todoItemRemoved } from 'src/features/todoItems/todoItemsSlice';

import {
    MUIColor
} from 'src/static/enums'

type Props = {
    itemId: string,
    color: MUIColor
}

function RemoveItemButtton({ itemId, color }: Props) {
    const dispatch = useAppDispatch();

    const onRemove = () => {
        dispatch(todoItemRemoved(itemId));
    }

    return (
        <IconButton
            id='remove-item-button'
            aria-label='Remove item button'
            onClick={onRemove}
            size='sm'
            variant='plain'
            color={color}
        >
            <Delete sx={{ color: 'inherit' }} />
        </IconButton>
    );
}

export default RemoveItemButtton;