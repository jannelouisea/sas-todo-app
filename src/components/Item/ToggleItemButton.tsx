import IconButton from '@mui/joy/IconButton';
import DoneIcon from '@mui/icons-material/Done';
import PropTypes from 'prop-types';

import {
    MUIColor,
    MUIVariant,
    MUISize
} from 'src/static/enums'

type Props = {
    color: MUIColor,
    variant: MUIVariant,
    onClick: React.MouseEventHandler<HTMLButtonElement>,
}

function ToggleItemButton({ color, variant, onClick }: Props) {
    return (
        <IconButton
            color={color}
            id='complete-item-button'
            variant={variant}
            aria-label='Complete item button'
            onClick={onClick}
            sx={{ borderRadius: '50%' }}
            size={MUISize.Small}
        >
            <DoneIcon fontSize='inherit' />
        </IconButton>
    );
}

ToggleItemButton.propTypes = {
    color: PropTypes.oneOf([...Object.values(MUIColor)]),
    variant: PropTypes.oneOf([...Object.values(MUIVariant)]),
    onClick: () => { }
}

ToggleItemButton.defaultProps = {
    color: MUIColor.Primary,
    variant: MUIVariant.Outlined,
    onClick: () => { }
}

export default ToggleItemButton;