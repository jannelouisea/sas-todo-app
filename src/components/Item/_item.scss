.to-do-item {
    display: flex;
    align-items: center;

    &_toggle {
        margin-left: .25rem;
        margin-right: 1rem;
    }

    &_text {
        flex: 1 1 0%;
    }

    &_remove-btn {
        margin-right: .25rem;
        margin-left: 1rem;
        visibility: hidden;
    }
}

.to-do-item:hover .to-do-item_remove-btn {
    visibility: visible;
}