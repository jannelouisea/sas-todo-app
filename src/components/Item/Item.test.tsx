import { render } from '@testing-library/react';
import Item from './Item';

test('renders component', () => {
    //render(<Item />);
    expect(true).toBeTruthy();
});
