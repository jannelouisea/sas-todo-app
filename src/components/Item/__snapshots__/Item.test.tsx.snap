// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders completed item 1`] = `
<div
  className="MuiCard-root MuiCard-vertical MuiCard-variantSoft MuiCard-colorSuccess MuiCard-sizeSm drop-shadow css-17iutky-JoyCard-root"
>
  <div
    className="to-do-item"
    data-first-child=""
    data-last-child=""
  >
    <div
      className="to-do-item_toggle"
    >
      <button
        aria-label="Complete item button"
        className="MuiIconButton-root MuiIconButton-variantSolid MuiIconButton-colorSuccess MuiIconButton-sizeSm css-qxb87h-JoyIconButton-root"
        id="complete-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="DoneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
    </div>
    <div
      className="to-do-item_text"
    >
      <span>
        To-do foo
      </span>
      <div
        className="item-created-date"
      >
        <p
          className="MuiTypography-root MuiTypography-body3 MuiTypography-colorNeutral MuiTypography-variantPlain css-1oc79or-JoyTypography-root"
        >
          Created Jul 15, 2023 12:00 AM
        </p>
        <div
          className="hidden"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
            data-testid="HourglassBottomIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="to-do-item_remove-btn"
    >
      <button
        aria-label="Remove item button"
        aria-labelledby={null}
        className="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorSuccess MuiIconButton-sizeSm css-1raploq-JoyIconButton-root"
        id="remove-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ekxpud-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Item renders correctly 1`] = `
<div
  className="MuiCard-root MuiCard-vertical MuiCard-variantOutlined MuiCard-colorWarning MuiCard-sizeSm drop-shadow css-1l37zdt-JoyCard-root"
>
  <div
    className="to-do-item"
    data-first-child=""
    data-last-child=""
  >
    <div
      className="to-do-item_toggle"
    >
      <button
        aria-label="Complete item button"
        className="MuiIconButton-root MuiIconButton-variantOutlined MuiIconButton-colorWarning MuiIconButton-sizeSm css-pxh0lg-JoyIconButton-root"
        id="complete-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="DoneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
    </div>
    <div
      className="to-do-item_text"
    >
      <span>
        To-do foo
      </span>
      <div
        className="item-created-date"
      >
        <p
          className="MuiTypography-root MuiTypography-body3 MuiTypography-colorNeutral MuiTypography-variantPlain css-gfqtt1-JoyTypography-root"
        >
          Created Jul 15, 2023 12:00 AM
        </p>
        <div
          className=""
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
            data-testid="HourglassBottomIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="to-do-item_remove-btn"
    >
      <button
        aria-label="Remove item button"
        aria-labelledby={null}
        className="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorWarning MuiIconButton-sizeSm css-punmis-JoyIconButton-root"
        id="remove-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ekxpud-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Item renders newer item 1`] = `
<div
  className="MuiCard-root MuiCard-vertical MuiCard-variantOutlined MuiCard-colorWarning MuiCard-sizeSm drop-shadow css-1l37zdt-JoyCard-root"
>
  <div
    className="to-do-item"
    data-first-child=""
    data-last-child=""
  >
    <div
      className="to-do-item_toggle"
    >
      <button
        aria-label="Complete item button"
        className="MuiIconButton-root MuiIconButton-variantOutlined MuiIconButton-colorWarning MuiIconButton-sizeSm css-pxh0lg-JoyIconButton-root"
        id="complete-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="DoneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
    </div>
    <div
      className="to-do-item_text"
    >
      <span>
        To-do foo
      </span>
      <div
        className="item-created-date"
      >
        <p
          className="MuiTypography-root MuiTypography-body3 MuiTypography-colorNeutral MuiTypography-variantPlain css-gfqtt1-JoyTypography-root"
        >
          Created Jun 15, 2023 12:00 AM
        </p>
        <div
          className=""
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
            data-testid="HourglassBottomIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="to-do-item_remove-btn"
    >
      <button
        aria-label="Remove item button"
        aria-labelledby={null}
        className="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorWarning MuiIconButton-sizeSm css-punmis-JoyIconButton-root"
        id="remove-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ekxpud-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;

exports[`Item renders older item 1`] = `
<div
  className="MuiCard-root MuiCard-vertical MuiCard-variantOutlined MuiCard-colorWarning MuiCard-sizeSm drop-shadow css-1l37zdt-JoyCard-root"
>
  <div
    className="to-do-item"
    data-first-child=""
    data-last-child=""
  >
    <div
      className="to-do-item_toggle"
    >
      <button
        aria-label="Complete item button"
        className="MuiIconButton-root MuiIconButton-variantOutlined MuiIconButton-colorWarning MuiIconButton-sizeSm css-pxh0lg-JoyIconButton-root"
        id="complete-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
          data-testid="DoneIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
    </div>
    <div
      className="to-do-item_text"
    >
      <span>
        To-do foo
      </span>
      <div
        className="item-created-date"
      >
        <p
          className="MuiTypography-root MuiTypography-body3 MuiTypography-colorNeutral MuiTypography-variantPlain css-gfqtt1-JoyTypography-root"
        >
          Created Jun 15, 2023 12:00 AM
        </p>
        <div
          className=""
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-17e890r-MuiSvgIcon-root"
            data-testid="HourglassBottomIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m18 22-.01-6L14 12l3.99-4.01L18 2H6v6l4 4-4 3.99V22h12zM8 7.5V4h8v3.5l-4 4-4-4z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      className="to-do-item_remove-btn"
    >
      <button
        aria-label="Remove item button"
        aria-labelledby={null}
        className="MuiIconButton-root MuiIconButton-variantPlain MuiIconButton-colorWarning MuiIconButton-sizeSm css-punmis-JoyIconButton-root"
        id="remove-item-button"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        <svg
          aria-hidden={true}
          className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1ekxpud-MuiSvgIcon-root"
          data-testid="DeleteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
